server/schedule/checker/priority_checker.go

```diff
@@ -0,0 +1,134 @@
+// PriorityChecker ensures high priority region should run first
+type PriorityChecker struct {
+	cluster opt.Cluster
+	opts    *config.PersistOptions
+	queue   *cache.PriorityQueue
+}
+
+// NewPriorityChecker creates a priority checker.
+func NewPriorityChecker(cluster opt.Cluster) *PriorityChecker {
+	return &PriorityChecker{
+		cluster: cluster,
+		opts:    cluster.GetOpts(),
+		queue:   cache.NewPriorityQueue(defaultPriorityQueueSize),
+	}
+}
+
+// GetType returns PriorityChecker's type
+func (p *PriorityChecker) GetType() string {
+	return "priority-checker"
+}
+
+// RegionPriorityEntry records region priority info
+type RegionPriorityEntry struct {
+	Retry    int
+	Last     time.Time
+	regionID uint64
+}
+
+// ID implement PriorityQueueItem interface
+func (r RegionPriorityEntry) ID() uint64 {
+	return r.regionID
+}
+
+// NewRegionEntry construct of region priority entry
+func NewRegionEntry(regionID uint64) *RegionPriorityEntry {
+	return &RegionPriorityEntry{regionID: regionID, Last: time.Now(), Retry: 1}
+}
+
+// Check check region's replicas, it will put into priority queue if the region lack of replicas.
+func (p *PriorityChecker) Check(region *core.RegionInfo) {
+	makeupCount := 0
+	if p.opts.IsPlacementRulesEnabled() {
+		makeupCount = p.checkRegionInPlacementRule(region)
+	} else {
+		makeupCount = p.checkRegionInReplica(region)
+	}
+	priority := 0 - makeupCount
+	p.addPriorityQueue(priority, region.GetID())
+}
+
+// checkRegionInPlacementRule check region in placement rule mode
+func (p *PriorityChecker) checkRegionInPlacementRule(region *core.RegionInfo) (makeupCount int) {
+	fit := p.cluster.FitRegion(region)
+	if len(fit.RuleFits) == 0 {
+		return
+	}
+
+	for _, rf := range fit.RuleFits {
+		// skip learn rule
+		if rf.Rule.Role == placement.Learner {
+			continue
+		}
+		makeupCount = makeupCount + rf.Rule.Count - len(rf.Peers)
+	}
+	return
+}
+
+// checkReplicas check region in replica mode
+func (p *PriorityChecker) checkRegionInReplica(region *core.RegionInfo) (makeupCount int) {
+	return p.opts.GetMaxReplicas() - len(region.GetPeers())
+}
+
+// addPriorityQueue add region into queue
+// it will remove if region's priority equal 0
+// it's retry will increase if region's priority equal last
+func (p *PriorityChecker) addPriorityQueue(priority int, regionID uint64) {
+	if priority < 0 {
+		if entry := p.queue.Get(regionID); entry != nil && entry.Priority == priority {
+			e := entry.Value.(*RegionPriorityEntry)
+			e.Retry = e.Retry + 1
+			e.Last = time.Now()
+		}
+		entry := NewRegionEntry(regionID)
+		p.queue.Put(priority, entry)
+	} else {
+		p.queue.Remove(regionID)
+	}
+}
+
+// GetPriorityRegions returns all regions in priority queue that needs rerun
+func (p *PriorityChecker) GetPriorityRegions() (ids []uint64) {
+	entries := p.queue.Elems()
+	for _, e := range entries {
+		re := e.Value.(*RegionPriorityEntry)
+		if t := re.Last.Add(time.Duration(re.Retry*10) * p.opts.GetPatrolRegionInterval()); t.Before(time.Now()) {
```

Comment:
Can you add a comment to explain it?

Reply:
the description as below: 
// avoid to some priority region occupy checker, region don't need check on next check interval
// the next run time is : last_time+retry*10*patrol_region_interval